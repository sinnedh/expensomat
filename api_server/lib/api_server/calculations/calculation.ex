defmodule ApiServer.Calculations.Calculation do
  use Ecto.Schema
  import Ecto.Changeset
  alias ApiServer.Calculations.Calculation


  schema "calculations" do
    field :description, :string
    field :name, :string
    has_many :members, ApiServer.Calculations.Member

    timestamps()
  end

  @doc false
  def changeset(%Calculation{} = calculation, attrs) do
    calculation
    |> cast(attrs, [:name, :description])
    |> validate_required([:name])
  end
end
