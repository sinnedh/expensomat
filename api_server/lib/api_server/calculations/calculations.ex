defmodule ApiServer.Calculations do
  @moduledoc """
  The Calculations context.
  """

  import Ecto.Query, warn: false
  alias ApiServer.Repo

  alias ApiServer.Calculations.Calculation

  @doc """
  Returns the list of calculations (optionally with preloaded members).

  ## Examples

      iex> list_calculations()
      [%Calculation{}, ...]

  """
  def list_calculations do
    Repo.all(Calculation)
  end
  def list_calculations_with_members do
    list_calculations
    |> Repo.preload(:members)
  end

  @doc """
  Gets a single calculation (optionally with preloaded members).

  Raises `Ecto.NoResultsError` if the Calculation does not exist.

  ## Examples

      iex> get_calculation!(123)
      %Calculation{}

      iex> get_calculation!(456)
      ** (Ecto.NoResultsError)

  """
  def get_calculation!(id), do: Repo.get!(Calculation, id)
  def get_calculation_with_members!(id), do: get_calculation!(id) |> Repo.preload(:members)

  @doc """
  Creates a calculation.

  ## Examples

      iex> create_calculation(%{field: value})
      {:ok, %Calculation{}}

      iex> create_calculation(%{field: bad_value})
      {:error, %Ecto.Changeset{}}

  """
  def create_calculation(attrs \\ %{}) do
    %Calculation{}
    |> Calculation.changeset(attrs)
    |> Repo.insert()
  end

  @doc """
  Updates a calculation.

  ## Examples

      iex> update_calculation(calculation, %{field: new_value})
      {:ok, %Calculation{}}

      iex> update_calculation(calculation, %{field: bad_value})
      {:error, %Ecto.Changeset{}}

  """
  def update_calculation(%Calculation{} = calculation, attrs) do
    calculation
    |> Calculation.changeset(attrs)
    |> Repo.update()
  end

  @doc """
  Deletes a Calculation.

  ## Examples

      iex> delete_calculation(calculation)
      {:ok, %Calculation{}}

      iex> delete_calculation(calculation)
      {:error, %Ecto.Changeset{}}

  """
  def delete_calculation(%Calculation{} = calculation) do
    Repo.delete(calculation)
  end

  @doc """
  Returns an `%Ecto.Changeset{}` for tracking calculation changes.

  ## Examples

      iex> change_calculation(calculation)
      %Ecto.Changeset{source: %Calculation{}}

  """
  def change_calculation(%Calculation{} = calculation) do
    Calculation.changeset(calculation, %{})
  end

  alias ApiServer.Calculations.Expense

  @doc """
  Returns the list of expenses for a given calculation.

  ## Examples

      iex> list_expenses(23)
      [%Expense{}, ...]

  """
  def list_expenses(calculation_id) do
    Repo.all(from e in Expense, where: e.calculation_id == ^calculation_id)
  end

  def list_expenses(calculation_id, :preload_members) do
    list_expenses(calculation_id)
    |> Repo.preload([:paid_by, :paid_for])
  end

  @doc """
  Returns the list of expenses.

  ## Examples

      iex> list_expenses()
      [%Expense{}, ...]

  """
  def list_expenses do
    Repo.all(Expense)
  end

  @doc """
  Gets a single expense.

  Raises `Ecto.NoResultsError` if the Expense does not exist.

  ## Examples

      iex> get_expense!(123)
      %Expense{}

      iex> get_expense!(456)
      ** (Ecto.NoResultsError)

  """
  def get_expense!(id) do
    Repo.get!(Expense, id)
  end

  def get_expense!(id, :preload_members) do
    get_expense!(id)
    |> Repo.preload([:paid_by, :paid_for])
  end

  @doc """
  Creates a expense.

  ## Examples

      iex> create_expense(%{field: value})
      {:ok, %Expense{}}

      iex> create_expense(%{field: bad_value})
      {:error, %Ecto.Changeset{}}

  """
  def create_expense(attrs \\ %{}) do
    %Expense{}
    |> Expense.changeset(attrs)
    |> Repo.insert()
  end

  @doc """
  Updates a expense.

  ## Examples

      iex> update_expense(expense, %{field: new_value})
      {:ok, %Expense{}}

      iex> update_expense(expense, %{field: bad_value})
      {:error, %Ecto.Changeset{}}

  """
  def update_expense(%Expense{} = expense, attrs) do
    expense
    |> Expense.changeset(attrs)
    |> Repo.update()
  end

  @doc """
  Deletes a Expense.

  ## Examples

      iex> delete_expense(expense)
      {:ok, %Expense{}}

      iex> delete_expense(expense)
      {:error, %Ecto.Changeset{}}

  """
  def delete_expense(%Expense{} = expense) do
    Repo.delete(expense)
  end

  @doc """
  Returns an `%Ecto.Changeset{}` for tracking expense changes.

  ## Examples

      iex> change_expense(expense)
      %Ecto.Changeset{source: %Expense{}}

  """
  def change_expense(%Expense{} = expense) do
    Expense.changeset(expense, %{})
  end

  alias ApiServer.Calculations.Member

  @doc """
  Returns the list of members.

  ## Examples

      iex> list_members()
      [%Member{}, ...]

  """
  def list_members do
    Repo.all(Member)
  end

  @doc """
  Gets a single member.

  Raises `Ecto.NoResultsError` if the Member does not exist.

  ## Examples

      iex> get_member!(123)
      %Member{}

      iex> get_member!(456)
      ** (Ecto.NoResultsError)

  """
  def get_member!(id), do: Repo.get!(Member, id)

  @doc """
  Creates a member.

  ## Examples

      iex> create_member(%{field: value})
      {:ok, %Member{}}

      iex> create_member(%{field: bad_value})
      {:error, %Ecto.Changeset{}}

  """
  def create_member(attrs \\ %{}) do
    %Member{}
    |> Member.changeset(attrs)
    |> Repo.insert()
  end

  @doc """
  Updates a member.

  ## Examples

      iex> update_member(member, %{field: new_value})
      {:ok, %Member{}}

      iex> update_member(member, %{field: bad_value})
      {:error, %Ecto.Changeset{}}

  """
  def update_member(%Member{} = member, attrs) do
    member
    |> Member.changeset(attrs)
    |> Repo.update()
  end

  @doc """
  Deletes a Member.

  ## Examples

      iex> delete_member(member)
      {:ok, %Member{}}

      iex> delete_member(member)
      {:error, %Ecto.Changeset{}}

  """
  def delete_member(%Member{} = member) do
    Repo.delete(member)
  end

  @doc """
  Returns an `%Ecto.Changeset{}` for tracking member changes.

  ## Examples

      iex> change_member(member)
      %Ecto.Changeset{source: %Member{}}

  """
  def change_member(%Member{} = member) do
    Member.changeset(member, %{})
  end
end
