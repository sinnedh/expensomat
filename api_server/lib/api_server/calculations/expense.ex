defmodule ApiServer.Calculations.Expense do
  use Ecto.Schema
  import Ecto.Changeset
  import Ecto.Query
  alias ApiServer.Calculations.{Expense, Member}
  alias ApiServer.Repo


  schema "expenses" do
    field :amount, :integer
    field :description, :string
    field :calculation_id, :id
    field :paid_at, :utc_datetime
    many_to_many :paid_by, ApiServer.Calculations.Member, join_through: "expense_paid_by_members"
    many_to_many :paid_for, ApiServer.Calculations.Member, join_through: "expense_paid_for_members"
    field :deleted_at, :utc_datetime

    timestamps()
  end

  def deleted(query) do
    query |> where([e], not is_nil(e.deleted_at))
  end

  def undeleted(query) do
    query |> where([e], is_nil(e.deleted_at))
  end

  def delete_changeset(%Expense{} = calculation) do
    current_time = DateTime.utc_now
    calculation
    |> cast(%{deleted_at: current_time}, [:deleted_at])
    |> validate_required([:deleted_at])
  end

  @doc false
  def changeset(%Expense{} = expense, attrs) do
    expense
    |> Repo.preload([:paid_by, :paid_for])
    |> cast(attrs, [:calculation_id, :description, :amount, :paid_at, :deleted_at])
    |> validate_required([:calculation_id, :description, :amount, :paid_at])
    |> put_paid_for(attrs)
    |> put_paid_by(attrs)
  end

  defp put_paid_for(changeset, %{"paid_for" => paid_for}) do
    members = Repo.all(from m in Member, where: m.id in ^paid_for)
    changeset |> put_assoc(:paid_for, members)
  end
  defp put_paid_for(changeset, _), do: changeset

  defp put_paid_by(changeset, %{"paid_by" => paid_by}) do
    members = Repo.all(from m in Member, where: m.id in ^paid_by)
    changeset |> put_assoc(:paid_by, members)
  end
  defp put_paid_by(changeset, _), do: changeset

end
