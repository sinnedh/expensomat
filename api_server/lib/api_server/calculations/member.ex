defmodule ApiServer.Calculations.Member do
  use Ecto.Schema
  import Ecto.Changeset
  alias ApiServer.Calculations.Member


  schema "members" do
    field :token, :string  # TODO: at least hash this token when being stored (e.g. using comeonin/bcrypt)
    field :name, :string
    field :calculation_id, :id

    timestamps()
  end

  @doc false
  def changeset(%Member{} = member, attrs) do
    member
    |> cast(attrs, [:name, :calculation_id])
    |> validate_required([:name])
    |> create_token_if_not_exists
    |> unique_constraint(:token)
  end

  defp create_token_if_not_exists(changeset) do
    token = get_field(changeset, :token)

    cond do
      token == nil -> changeset |> put_change(:token, generate_token())
      true -> changeset
    end
  end

  def generate_token() do
    length = 24
    :crypto.strong_rand_bytes(length)
    |> Base.url_encode64
    |> binary_part(0, length)
  end
end
