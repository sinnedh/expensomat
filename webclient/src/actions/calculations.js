import {fetchCalculation} from '../api';

export const getCalculation = (token) => {
  return dispatch => {
    dispatch(requestCalculation());
    return fetchCalculation(
      dispatch,
      token,
      receiveCalculationSuccess,
      receiveCalculationFailure,
    )
  }
}

export const requestCalculation = () => {
  return {type: 'CALCULATION:LOAD_REQUEST'}
}

export const receiveCalculationSuccess = (data) => {
  return {
    type: 'CALCULATION:LOAD_SUCCESS',
    name: data.name,
    description: data.description,
    members: data.members,
    matrix: data.matrix,
  }
}

export const receiveCalculationFailure = (error) => {
  return {type: 'CALCULATION:LOAD_FAILURE', message: error.message}
}
