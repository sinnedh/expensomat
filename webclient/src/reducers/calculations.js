import { List, Map } from 'immutable'

const initialState = Map({
  name: '',
  description: '',
  members: List([]),
  matrix: Map({}),
})

const calculations = (state = initialState, action) => {
  switch (action.type) {
    case 'CALCULATION:LOAD_SUCCESS':
      return state.merge({
        name: action.name,
        description: action.description,
        members: action.members,
        matrix: action.matrix,
      })


    case 'CALCULATION:CREATE_SUCCESS':
      return state.merge({
        name: action.name,
        description: action.description,
        members: action.members,
      });

    case 'CALCULATION:DELETE_SUCCESS':
      return initialState;

    case 'CALCULATION:UPDATE_SUCCESS':
      return state.merge({
        name: action.name,
        description: action.description,
        members: action.members,
      });

    case 'CALCULATION:LOAD_REQUEST':
    case 'CALCULATION:LOAD_FAILURE':
    case 'CALCULATION:CREATE_REQUEST':
    case 'CALCULATION:CREATE_FAILURE':
    case 'CALCULATION:DELETE_REQUEST':
    case 'CALCULATION:DELETE_FAILURE':
    case 'CALCULATION:UPDATE_REQUEST':
    case 'CALCULATION:UPDATE_FAILURE':
    default:
      return state;
  }
}

export default calculations;
