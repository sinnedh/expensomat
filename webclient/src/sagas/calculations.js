import { call, put } from 'redux-saga/effects'
import { setErrorNotification, setInfoNotification } from '../actions'
import { incrementLoadingCounter, decrementLoadingCounter } from '../actions'
import * as api from '../api'

export function* createCalculation(action) {
  yield put(incrementLoadingCounter())
  try {
    const calculation = yield call(api.createCalculation, action.token, action.calculation)
    yield put({
      type: "CALCULATION:CREATE_SUCCESS",
      name: calculation.data.name,
      members: calculation.data.members,
      description: calculation.data.description,
    })
    yield put(setInfoNotification('Calculation created succesfully'))
  } catch (e) {
    yield put({type: "CALCULATION:CREATE_FAILURE", message: e.message})
    yield put(setErrorNotification(`Could not create calculation ("${e.message}")`))
  } finally {
    yield put(decrementLoadingCounter())
  }
}

export function* fetchCalculation(action) {
  yield put(incrementLoadingCounter())

  try {
    const calculation = yield call(api.fetchCalculation, action.token)

    yield put({
      type: "CALCULATION:LOAD_SUCCESS",
      name: calculation.data.name,
      members: calculation.data.members,
      description: calculation.data.description,
      matrix: calculation.data.matrix,
    })
    yield put(setInfoNotification('Calculation loaded succesfully'))
  } catch (e) {
    yield put({type: "CALCULATION:LOAD_FAILURE", message: e.message})
    yield put(setErrorNotification(`Could not load calculation ("${e.message}")`))
  } finally {
    yield put(decrementLoadingCounter())
  }
}
