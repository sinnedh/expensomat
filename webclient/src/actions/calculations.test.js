import * as actions from './calculations'

describe('CALCULATIONS', () => {
  const token = 'ABCD123'
  const calculation = {id: 1, name: 'Calculation'}

  it('handles getCalculation', () => {
    const action = actions.getCalculation(token)
    expect(action).toMatchSnapshot()
  })

  it('handles createCalculation', () => {
    const action = actions.createCalculation(calculation)
    expect(action).toMatchSnapshot()
  })

  it('handles deleteCalculation', () => {
    const action = actions.deleteCalculation(token, calculation)
    expect(action).toMatchSnapshot()
  })

  it('handles updateCalculation', () => {
    const action = actions.updateCalculation(token, {name: 'Calculation'})
    expect(action).toMatchSnapshot()
  })

  it('handles updateMember', () => {
    const members = [{id: 1, name: 'Horst'}, {id: 2, name: 'Dirk'}, {id: 3, name: 'Kalle'}]
    const member = {id: 2, name: 'Dirk'}
    const changes = {name: 'Keek'}
    const action = actions.updateMember(token, members, member, changes)
    expect(action).toMatchSnapshot()
  })
})
