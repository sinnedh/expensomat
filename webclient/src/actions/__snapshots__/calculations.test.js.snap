// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CALCULATIONS handles createCalculation 1`] = `
Object {
  "calculation": Object {
    "id": 1,
    "name": "Calculation",
  },
  "type": "CALCULATION:CREATE_REQUEST",
}
`;

exports[`CALCULATIONS handles deleteCalculation 1`] = `
Object {
  "calculation": Object {
    "id": 1,
    "name": "Calculation",
  },
  "token": "ABCD123",
  "type": "CALCULATION:DELETE_REQUEST",
}
`;

exports[`CALCULATIONS handles getCalculation 1`] = `
Object {
  "token": "ABCD123",
  "type": "CALCULATION:LOAD_REQUEST",
}
`;

exports[`CALCULATIONS handles updateCalculation 1`] = `
Object {
  "changes": Object {
    "name": "Calculation",
  },
  "token": "ABCD123",
  "type": "CALCULATION:UPDATE_REQUEST",
}
`;
